### ISSUE TITLE
Tweak SK_GAMMA_EXPONENT and SK_GAMMA_CONTRAST to get proper gamma correction for text

### ISSUE THREAD

**Issue by Yasand123** [ungoogled-software/ungoogled-chromium-archlinux#212]
[Background](https://bugs.chromium.org/p/chromium/issues/detail?id=882264).

You are probably familiar with this unfortunate limitation we have on Linux. Thankfully there's an option to apply proper gamma levels for text on Chromium independently, but it is not tweaked by default for the GNU/Linux operating system. It would be nice to have this done by you guys.

**Comment by jstkdng** [comment]
So, you'd like us to add a patch that changes the gamma values? to which values exactly?

**Comment by Yasand123** [comment]
> So, you'd like us to add a patch that changes the gamma values? to which values exactly?

General consensus is that it should be somewhere between 1.8 and 2.2. This [test](https://bel.fi/alankila/lcd/gamma.html) by Alankila should give us a good idea on what the value should be. From my testing on different displays, I find 2.2 is the closest to what I would consider a good default. In reality there's no "correct" exponent value, but 2.2 seems to be a very good approximation and any error caused by it is minute and certainly better than having no gamma correction at all.

That's for `SK_GAMMA_EXPONENT`. As for `SK_GAMMA_CONTRAST` maybe it should be ignored for now because it's more of a preference.

**Comment by jstkdng** [comment]
ok, the patch seems simple enough to maintain
```
diff --git a/skia/BUILD.gn b/skia/BUILD.gn
index b2c8b79..7e2ee53 100644
--- a/skia/BUILD.gn
+++ b/skia/BUILD.gn
@@ -139,7 +139,7 @@ config("skia_library_config") {
   # Settings for text blitting, chosen to approximate the system browser.
   if (is_linux || is_chromeos) {
     defines += [
-      "SK_GAMMA_EXPONENT=1.2",
+      "SK_GAMMA_EXPONENT=2.2",
       "SK_GAMMA_CONTRAST=0.2",
     ]
   } else if (is_android) {
```

**Comment by jstkdng** [comment]
Ok, I've built a new version with that patch enabled.
https://download.opensuse.org/repositories/home:/justkidding:/arch/Arch/x86_64/ungoogled-chromium-109.0.5414.74-2-x86_64.pkg.tar.zst

can you give it a test?

**Comment by Yasand123** [comment]
Alright. After some testing it seems 2.2 for `SK_GAMMA_EXPONENT` is good. Contrast might need to be increased a little, maybe to ~0.5, but that needs more testing. Correcting the gamma level fixed the semi-broken rendering for dark themes in websites, as previously the dark background would bleed into the text affecting its thickness and color.

As for light themes, that's a different story. Increasing the gamma exponent meant, by definition, that the dark text would look a little too thin on lighter backgrounds. You either get proper dark mode or proper light mode but not both simultaneously (by default).

This issue was expected, and I am currently playing with some [stem darkening parameters](https://freetype.org/freetype2/docs/reference/ft2-properties.html#no-stem-darkening) for FreeType. Stem darkening was created as a counter measure for exactly this purpose, and both Windows and macOS use it as both do gamma corrected rendering system-wide.

The settings in question are `autofitter:darkening-parameters` and `cff:darkening-parameters`.

Unfortunately I can't seem to be able get the proper values yet, I had the impression that it would be a lot more straightforward but seeing how time consuming this process is, I doubt the average user would bother. The good news is that if you're primarily a dark mode user you'll be happy without having to tweak anything.

I'll be testing stuff from time to time and I'll make sure to give an update if I ever find a good solution. But yeah, we definitely cannot make this the default unfortunately.

Maybe keep the link and patch here for those who are interested. I personally will probably keep patching ungoogled-chromium as I'm primarily a dark mode user.

Thank you for your time, and for the pre-compiled package, I really appreciate it! If you have no further comments feel free to close the issue.

**Comment by Yasand123** [comment]
Update: I did one last test and I think this looks very acceptable now, only enabling stem darkening without editing any of the extra parameters (so `cff:no-stem-darkening=0` and `autofitter:no-stem-darkening=0` exported as env variables) gives us a good result for light themes. So I guess it's still a viable option for those who want it? The issue is, I am not sure if everyone will be happy about the changes, and I don't think you can maintain 2 builds simultaneously. It's ultimately up to you.

**Comment by taprobane99** [comment]
Is there any update on this? As a Mac user it was immediately obvious to me that text in Linux on chromium is too gray/faint. Firefox renders it with the correct gamma.

**Comment by iGerman00** [comment]
Bumping this, I have been on a journey to try and debug this and ended up finally succeeding in compiling my own test build. This is a useful and needed addition, I detailed more and shared screenshot comparisons on the Vivaldi forums:
https://forum.vivaldi.net/post/849737

**Comment by PF4Public** [comment]
> Also, I just wanted to bump the issue because not a single soul cares about proper font gamma on Chromium elsewhere, when it’s a tiny change.

Perhaps it is your setup that is somehow different? That's how it looks for me with dark theme and without any changes to gamma:

<img width="759" height="513" alt="Image" src="https://github.com/user-attachments/assets/41c42f33-f90b-4a77-9f9a-345010e086f0" />

Looks OK, doesn't it? It might look blurry on a screenshot, but I guess it is due to hidpi-screen.

**Comment by iGerman00** [comment]
Yeah, you're right, setup differences are a thing. It will look okay on a HiDPI screen no matter what because it's far less applicable there, no matter the gamma - you can set it to 1.0 and it will just be really thin fonts in dark mode and really thick ones in light mode, but probably still readable. It makes sense: you get more resolution and bigger letters to span it - I can tell you've got scaling on in your system.

Normal screens, which is what most people have (even if they're relatively high-resolution like 3440x1440 @ 34" in my case), suffer a *lot* from bad font gamma, so no, it's not just my setup, it's *just yours* haha. When text gets small, it has to be antialiased and go in-between a small number of pixels, and antialiasing is where gamma comes in strong. Of course, if I zoom in, the issue gets exponentially less noticeable, but nobody browses every web page at 200% scaling.

I consistently struggle to read dark mode pages on all browsers I've tried, and this issue is specific to Linux because Chromium set an unreasonably low gamma on it for no real reason. It also annoys me greatly that this is very platform specific - nothing like this is present on Mac, Windows or even mobile OSes. This is one of the last little roadblocks that prevents me from enjoying desktop Linux :(. Either way, the default 1.2 and 0.2 values are *egregiously* incorrect, and I shame chromium every chance I get for it, because they also completely ignore reports about it.

---

I'm currently building a version to match Windows, with gamma exponent @ 2.2 and contrast at 1.0. This will fail pure 2.2 [font gamma tests](https://fontgammatest.replit.app), but probably subjectively look a bit better, as just setting the gamma makes fonts a little too thick when their weight is >400.
